flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala

```diff
@@ -342,6 +342,14 @@ object CodeGenUtils {
         throw new IllegalArgumentException("Illegal type: " + t)
     }
 
+  def castToPrecisionAndScale(fieldValTerm: String, to: LogicalType, from: LogicalType): String = {
+    (to, from) match {
+      case (toDec: DecimalType, _: DecimalType) =>
```

Comment:
This is a bit too specific. We should simply cast all arguments to `arrayType.getElementType` using the `generateCast` method.